src/Language/PureScript/TypeChecker/Deriving.hs

```diff
@@ -368,64 +498,118 @@ deriveFunctor
-          rebuilt = foldl' App ctor args
-          caseBinder = mkCtorBinder mn ctorName $ map mkBinder idents
-      return $ CaseAlternative [caseBinder] (unguarded rebuilt)
-      where
-        fVar = mkVar f
-        mapVar = mkRef Prelude.identMap
-
-        transformArg :: Ident -> SourceType -> m Expr
-        transformArg ident = fmap (foldr App (mkVar ident)) . goType where
-
-          goType :: SourceType -> m (Maybe Expr)
-          -- argument matches the index type
-          goType (TypeVar _ t) | t == iTyName = return (Just fVar)
+  mapVar = mkRef Prelude.identMap
 
-          -- records
-          goType recTy | Just row <- objectType recTy =
-              traverse buildUpdate (decomposeRec' row) >>= (traverse buildRecord . justUpdates)
-            where
-              justUpdates :: [Maybe (Label, Expr)] -> Maybe [(Label, Expr)]
-              justUpdates = foldMap (fmap return)
+toConst :: forall f a b. Functor f => f a -> Const [f a] b
+toConst = Const . pure
 
-              buildUpdate :: (Label, SourceType) -> m (Maybe (Label, Expr))
-              buildUpdate (lbl, ty) = do upd <- goType ty
-                                         return ((lbl,) <$> upd)
+consumeConst :: forall f a b c. Applicative f => ([a] -> b) -> Const [f a] c -> f b
+consumeConst f = fmap f . sequenceA . getConst
 
-              buildRecord :: [(Label, Expr)] -> m Expr
-              buildRecord updates = do
-                arg <- freshIdent "o"
-                let argVar = mkVar arg
-                    mkAssignment (Label l, x) = (l, App x (Accessor l argVar))
-                return (lam arg (ObjectUpdate argVar (mkAssignment <$> updates)))
+applyWhen :: forall a. Bool -> (a -> a) -> a -> a
+applyWhen cond f = if cond then f else identity
 
-          -- quantifiers
-          goType (ForAll _ scopedVar _ t _) | scopedVar /= iTyName = goType t
+deriveFoldable
+  :: forall m
+   . MonadError MultipleErrors m
+  => MonadState CheckState m
+  => MonadSupply m
+  => ModuleName
+  -> ProperName 'TypeName
+  -> m [(PSString, Expr)]
+deriveFoldable mn tyConNm = do
+  ctors <- validateParamsInTypeConstructors mn tyConNm
+  foldlFun <- mkAsymmetricFoldFunction False foldlVar ctors
+  foldrFun <- mkAsymmetricFoldFunction True foldrVar ctors
+  foldMapFun <- mkTraversal mn foldMapVar foldMapOps ctors
+  pure [(Foldable.foldl, foldlFun), (Foldable.foldr, foldrFun), (Foldable.foldMap, foldMapFun)]
+  where
+  foldlVar = mkRef Foldable.identFoldl
+  foldrVar = mkRef Foldable.identFoldr
+  foldMapVar = mkRef Foldable.identFoldMap
+  flipVar = mkRef Prelude.identFlip
+
+  mkAsymmetricFoldFunction :: Bool -> Expr -> [(ProperName 'ConstructorName, [Maybe ParamUsage])] -> m Expr
+  mkAsymmetricFoldFunction isRightFold recurseVar ctors = do
+    f <- freshIdent "f"
+    z <- freshIdent "z"
+    let
+      appCombiner :: (Bool, Expr) -> Expr -> Expr -> Expr
+      appCombiner (isFlipped, fn) = applyWhen (isFlipped == isRightFold) flip $ App . App fn
```

Comment:
I had a bit of a hard time understanding how `flip` works here. I'm writing out the code below to help clarify it for myself.
```haskell
-- assume `isFlipped == isRightFold` is `False`
appCombiner (_, fn) expr1 expr2 = App (App fn expr1) expr2

-- assume `isFlipped == isRightFold` is `True`
appCombiner (_, fn) expr1 expr2 = App expr2 (App fn expr1)
```

Reply:
Oh, not quite. The second case should be `App (App fn expr2) expr1`. `App . App fn` is a (Haskell) function that accepts two arguments and makes them the arguments to the PureScript function `fn`; flipping the Haskell function is thus equivalent to flipping the PureScript function.